man/backend.7
.\"
.\" Backend man page for CUPS.
.\"
.\" Copyright © 2020-2024 by OpenPrinting.
.\" Copyright © 2007-2019 by Apple Inc.
.\" Copyright © 1997-2006 by Easy Software Products.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH backend 7 "CUPS" "2021-02-28" "OpenPrinting"
.SH NAME
backend \- cups backend transmission interfaces
.SH SYNOPSIS
.B backend
.br
.B backend
.I job
.I user
.I title
.I num-copies
.I options
[
.I filename
]
.nf

\fB#include <cups/cups.h>\fR

\fBconst char *cupsBackendDeviceURI\fR(\fBchar **\fIargv\fR);

\fBvoid cupsBackendReport\fR(\fBconst char *\fIdevice_scheme\fR,
                       \fBconst char *\fIdevice_uri\fR,
                       \fBconst char *\fIdevice_make_and_model\fR,
                       \fBconst char *\fIdevice_info\fR,
                       \fBconst char *\fIdevice_id\fR,
                       \fBconst char *\fIdevice_location\fR);

\fBssize_t cupsBackChannelWrite\fR(\fBconst char *\fIbuffer\fR,
                             \fBsize_t \fIbytes\fR, \fBdouble \fItimeout\fR);

\fBint cupsSideChannelRead\fR(\fBcups_sc_command_t *\fIcommand\fR,
                        \fBcups_sc_status_t *\fIstatus\fR, \fBchar *\fIdata\fR,
                        \fBint *\fIdatalen\fR, \fBdouble \fItimeout\fR);

\fBint cupsSideChannelWrite\fR(\fBcups_sc_command_t \fIcommand\fR,
                         \fBcups_sc_status_t \fIstatus\fR, \fBconst char *\fIdata\fR,
                         \fBint \fIdatalen\fR, \fBdouble \fItimeout\fR);
.fi
.SH DESCRIPTION
Backends are a special type of
.BR filter (7)
which is used to send print data to and discover different devices on the system.
.LP
Like filters, backends must be capable of reading from a filename on the command-line or from the standard input, copying the standard input to a temporary file as required by the physical interface.
.LP
The command name (\fIargv[0]\fR) is set to the device URI of the destination printer.
Authentication information in
.I argv[0]
is removed, so backend developers are urged to use the
.B DEVICE_URI
environment variable whenever authentication information is required. The
.BR cupsBackendDeviceURI ()
function may be used to retrieve the correct device URI.
.LP
Back-channel data from the device should be relayed to the job filters using the \fIcupsBackChannelWrite\fR function.
.LP
Backends are responsible for reading side-channel requests using the
.BR cupsSideChannelRead ()
function and responding with the
.BR cupsSideChannelWrite ()
function. The
.B CUPS_SC_FD
constant defines the file descriptor that should be monitored for incoming requests.
.SS DEVICE DISCOVERY
When run with no arguments, the backend should list the devices and schemes it supports or is advertising to the standard output.
The output consists of zero or more lines consisting of any of the following forms:
.nf

    device-class scheme "Unknown" "device-info"
    device-class device-uri "device-make-and-model" "device-info"
    device-class device-uri "device-make-and-model" "device-info" "device-id"
    device-class device-uri "device-make-and-model" "device-info" "device-id" "device-location"
.fi
.LP
The
.BR cupsBackendReport ()
function can be used to generate these lines and handle any necessary escaping of characters in the various strings.
.LP
The
.I device-class
field is one of the following values:
.TP 5
.B direct
The device-uri refers to a specific direct-access device with no options, such as a parallel, USB, or SCSI device.
.TP 5
.B file
The device-uri refers to a file on disk.
.TP 5
.B network
The device-uri refers to a networked device and conforms to the general form for
network URIs.
.TP 5
.B serial
The device-uri refers to a serial device with configurable baud rate and other options.
If the device-uri contains a baud value, it represents the maximum baud rate supported by the device.
.LP
The
.I scheme
field provides the URI scheme that is supported by the backend.
Backends should use this form only when the backend supports any URI using that scheme.
The
.I device-uri
field specifies the full URI to use when communicating with the device.
.LP
The
.I device-make-and-model
field specifies the make and model of the device, e.g. "Example Foojet 2000".
If the make and model is not known, you must report "Unknown".
.LP
The
.I device-info
field specifies additional information about the device.
Typically this includes the make and model along with the port number or network address, e.g. "Example Foojet 2000 USB #1".
.LP
The optional
.I device-id
field specifies the IEEE-1284 device ID string for the device, which is used to select a matching driver.
.LP
The optional
.I device-location
field specifies the physical location of the device, which is often used to pre-populate the printer-location attribute when adding a printer.
.SS PERMISSIONS
Backends without world read and execute permissions are run as the root user.
Otherwise, the backend is run using an unprivileged user account, typically "lp".
.SH EXIT STATUS
The following exit codes are defined for backends:
.TP 5
.B CUPS_BACKEND_OK
The print file was successfully transmitted to the device or remote server.
.TP 5
.B CUPS_BACKEND_FAILED
.br
The print file was not successfully transmitted to the device or remote server.
The scheduler will respond to this by canceling the job, retrying the job, or stopping the queue depending on the state of the
.I printer-error-policy
attribute.
.TP 5
.B CUPS_BACKEND_AUTH_REQUIRED
The print file was not successfully transmitted because valid authentication information is required.
The scheduler will respond to this by holding the job and adding the 'cups-held-for-authentication' keyword to the "job-reasons" Job Description attribute.
.TP 5
.B CUPS_BACKEND_HOLD
The print file was not successfully transmitted because it cannot be printed at this time.
The scheduler will respond to this by holding the job.
.TP 5
.B CUPS_BACKEND_STOP
The print file was not successfully transmitted because it cannot be printed at this time.
The scheduler will respond to this by stopping the queue.
.TP 5
.B CUPS_BACKEND_CANCEL
The print file was not successfully transmitted because one or more attributes are not supported or the job was canceled at the printer.
The scheduler will respond to this by canceling the job.
.TP 5
.B CUPS_BACKEND_RETRY
The print file was not successfully transmitted because of a temporary issue.
The scheduler will retry the job at a future time - other jobs may print before this one.
.TP 5
.B CUPS_BACKEND_RETRY_CURRENT
The print file was not successfully transmitted because of a temporary issue.
The scheduler will retry the job immediately without allowing intervening jobs.
.PP
All other exit code values are reserved.
.SH ENVIRONMENT
In addition to the environment variables listed in
.BR cups (1)
and
.BR filter (7),
CUPS backends can expect the following environment variable:
.TP 5
.B DEVICE_URI
The device URI associated with the printer.
.SH FILES
.I /etc/cups/cups-files.conf
.SH NOTES
CUPS backends are not generally designed to be run directly by the user.
Aside from the device URI issue (
.I argv[0]
and
.B DEVICE_URI
environment variable contain the device URI), CUPS backends also expect specific environment variables and file descriptors, and typically run in a user session that (on macOS) has additional restrictions that affect how it runs.
Backends can also be installed with restricted permissions (0500 or 0700) that tell the scheduler to run them as the "root" user instead of an unprivileged user (typically "lp") on the system.
.LP
Unless you are a developer and know what you are doing, please do not run backends directly.
Instead, use the
.BR lp (1)
or
.BR lpr (1)
programs to send print jobs or
.BR lpinfo (8)
to query for available printers using the backend.
The one exception is the SNMP backend - see
.BR cups-snmp (8)
for more information.
.SH NOTES
CUPS printer drivers and backends are deprecated and will no longer be supported in a future feature release of CUPS.
Printers that do not support IPP can be supported using applications such as
.BR ippeveprinter (1).
.SH SEE ALSO
.IR cups (1),
.IR cups-files.conf (5),
.IR cups-snmp (8),
.IR cupsd (8),
.IR filter (7),
.IR lp (1),
.IR lpinfo (8),
.IR lpr (1),
.br
CUPS Online Help (http://localhost:631/help)
.SH COPYRIGHT
Copyright \[co] 2020-2024 by OpenPrinting.


CHANGES.md
CHANGES - OpenPrinting CUPS
===========================


Changes in CUPS v2.4.12 (YYYY-MM-DD)
------------------------------------

- GnuTLS follows system crypto policies now (Issue #1105)
- Added `NoSystem` SSLOptions value (Issue #1130)
- Fixed a compressed file error handling bug (Issue #1070)
- Fixed a bug in the make-and-model whitespace trimming code (Issue #1096)
- Fixed a removal of IPP Everywhere permanent queue if installation failed (Issue #1102)
- Fixed `ServerToken None` in scheduler (Issue #1111)
- Fixed the default User-Agent string.
- Fixed a recursion issue in `ippReadIO`.


Changes in CUPS v2.4.11 (2024-09-30)
------------------------------------

- Updated the maximum file descriptor limit for `cupsd` to 64k-1 (Issue #989)
- Fixed `lpoptions -d` with a discovered but not added printer (Issue #833)
- Fixed incorrect error message for HTTP/IPP errors (Issue #893)
- Fixed JobPrivateAccess and SubscriptionPrivateAccess support for "all"
  (Issue #990)
- Fixed issues with cupsGetDestMediaByXxx (Issue #993)
- Fixed adding and modifying of printers via the web interface (Issue #998)
- Fixed HTTP PeerCred authentication for domain users (Issue #1001)
- Fixed checkbox support (Issue #1008)
- Fixed printer state notifications (Issue #1013)
- Fixed IPP Everywhere printer setup (Issue #1033)


Changes in CUPS v2.4.10 (2024-06-18)
------------------------------------

- Fixed error handling when reading a mixed `1setOf` attribute.
- Fixed scheduler start if there is only domain socket to listen on (Issue #985)


Changes in CUPS v2.4.9 (2024-06-11)
-----------------------------------

- Fixed domain socket handling (CVE-2024-35235)
- Fixed creating of `cupsUrfSupported` PPD keyword (Issue #952)
- Fixed searching for destinations in web ui (Issue #954)
- Fixed TLS negotiation using OpenSSL with servers that require the TLS SNI
  extension.
- Really raised `cups_enum_dests()` timeout for listing available IPP printers
  (Issue #751)...
- Fixed `Host` header regression (Issue #967)
- Fixed DNS-SD lookups of local services with Avahi (Issue #970)
- Fixed listing jobs in destinations in web ui. (Apple issue #6204)
- Fixed showing search query in web ui help page. (Issue #977)


Changes in CUPS v2.4.8 (2024-04-26)
-----------------------------------

- Added warning if the device has to be asked for 'all,media-col-database'
  separately (Issue #829)
- Added new value for 'lpstat' option '-W' - successfull - for getting
  successfully printed jobs (Issue #830)
- Added support for PAM modules password-auth and system-auth (Issue #892)
- Updated IPP Everywhere printer creation error reporting (Issue #347)
- Updated and documented the MIME typing buffering limit (Issue #925)
- Now report an error for temporary printer defaults with lpadmin (Issue #237)
- Fixed mapping of PPD InputSlot, MediaType, and OutputBin values (Issue #238)
- Fixed "document-unprintable-error" handling (Issue #391)
- Fixed the web interface not showing an error for a non-existent printer
  (Issue #423)
- Fixed printing of jobs with job name longer than 255 chars on older printers
  (Issue #644)
- Really backported fix for Issue #742
- Fixed `cupsCopyDestInfo` device connection detection (Issue #586)
- Fixed "Upgrade" header handling when there is no TLS support (Issue #775)
- Fixed memory leak when unloading a job (Issue #813)
- Fixed memory leak when creating color profiles (Issue #815)
- Fixed a punch finishing bug in the IPP Everywhere support (Issue #821)
- Fixed crash in `scan_ps()` if incoming argument is NULL (Issue #831)
- Fixed setting job state reasons for successful jobs (Issue #832)
- Fixed infinite loop in IPP backend if hostname is IP address with Kerberos
  (Issue #838)
- Added additional check on socket if `revents` from `poll()` returns POLLHUP
  together with POLLIN or POLLOUT in `httpAddrConnect2()` (Issue #839)
- Fixed crash in `ppdEmitString()` if `size` is NULL (Issue #850)
- Fixed reporting `media-source-supported` when sharing printer which has
  numbers as strings instead of keywords as `InputSlot` values (Issue #859)
- Fixed IPP backend to support the "print-scaling" option with IPP printers
  (Issue #862)
- Fixed potential race condition for the creation of temporary queues
  (Issue #871)
- Fixed `httpGets` timeout handling (Issue #879)
- Fixed checking for required attributes during PPD generation (Issue #890)
- Fixed encoding of IPv6 addresses in HTTP requests (Issue #903)
- Fixed sending response headers to client (Issue #927)
- Fixed CGI program initialization and validation of form checkbox and text
  fields.


Changes in CUPS v2.4.7 (2023-09-20)
-----------------------------------

- CVE-2023-4504 - Fixed Heap-based buffer overflow when reading Postscript
  in PPD files
- Added OpenSSL support for cupsHashData (Issue #762)
- Fixed delays in lpd backend (Issue #741)
- Fixed extensive logging in scheduler (Issue #604)
- Fixed hanging of `lpstat` on IBM AIX (Issue #773)
- Fixed hanging of `lpstat` on Solaris (Issue #156)
- Fixed printing to stderr if we can't open cups-files.conf (Issue #777)
- Fixed purging job files via `cancel -x` (Issue #742)
- Fixed RFC 1179 port reserving behavior in LPD backend (Issue #743)
- Fixed a bug in the PPD command interpretation code (Issue #768)
- Fixed Oki 407 freeze when printing larger jobs (Issue #877)


Changes in CUPS v2.4.6 (2023-06-22)
-----------------------------------

- CVE-2023-34241: Fixed use-after-free when logging warnings in case of failures
  in `cupsdAcceptClient()`.
- Fixed linking error on old MacOS (Issue #715)
- Fixed printing multiple files on specific printers (Issue #643)


Changes in CUPS v2.4.5 (2023-06-13)
-----------------------------------

- Fixed corruption of locally saved certificates (Issue #724)


Changes in CUPS v2.4.4 (2023-06-06)
-----------------------------------

- Fixed segfault in `cupsGetNamedDest()` when trying to get default printer, but
  the default printer is not set (Issue #719)



cups/backend.c
/*
 * Backend functions for CUPS.
 *
 * Copyright © 2020-2025 by OpenPrinting.
 * Copyright 2007-2015 by Apple Inc.
 * Copyright 2006 by Easy Software Products.
 *
 * Licensed under Apache License v2.0.  See the file "LICENSE" for more information.
 */

/*
 * Include necessary headers...
 */

#include "cups-private.h"
#include "backend.h"
#include "ppd.h"


/*
 * Local functions...
 */

static void	quote_string(const char *s);


/*
 * 'cupsBackendDeviceURI()' - Get the device URI for a backend.
 *
 * The "argv" argument is the argv argument passed to main(). This
 * function returns the device URI passed in the DEVICE_URI environment
 * variable or the device URI passed in argv[0], whichever is found
 * first.
 *
 * @since CUPS 1.2/macOS 10.5@
 */

const char *				/* O - Device URI or @code NULL@ */
cupsBackendDeviceURI(char **argv)	/* I - Command-line arguments */
{
  const char	*device_uri,		/* Device URI */
		*auth_info_required;	/* AUTH_INFO_REQUIRED env var */
  _cups_globals_t *cg = _cupsGlobals();	/* Global info */
  int		options;		/* Resolve options */
  ppd_file_t	*ppd;			/* PPD file */
  ppd_attr_t	*ppdattr;		/* PPD attribute */


  if ((device_uri = getenv("DEVICE_URI")) == NULL)
  {
    if (!argv || !argv[0] || !strchr(argv[0], ':'))
      return (NULL);

    device_uri = argv[0];
  }

  options = _HTTP_RESOLVE_STDERR;
  if ((auth_info_required = getenv("AUTH_INFO_REQUIRED")) != NULL &&
      !strcmp(auth_info_required, "negotiate"))
    options |= _HTTP_RESOLVE_FQDN;

  if ((ppd = ppdOpenFile(getenv("PPD"))) != NULL)
  {
    if ((ppdattr = ppdFindAttr(ppd, "cupsIPPFaxOut", NULL)) != NULL &&
        !_cups_strcasecmp(ppdattr->value, "true"))
      options |= _HTTP_RESOLVE_FAXOUT;

    ppdClose(ppd);
  }

  return (_httpResolveURI(device_uri, cg->resolved_uri,
                          sizeof(cg->resolved_uri), options, NULL, NULL));
}


/*
 * 'cupsBackendReport()' - Write a device line from a backend.
 *
 * This function writes a single device line to stdout for a backend.
 * It handles quoting of special characters in the device-make-and-model,
 * device-info, device-id, and device-location strings.
 *
 * @since CUPS 1.4/macOS 10.6@
 */

void
cupsBackendReport(
    const char *device_scheme,		/* I - device-scheme string */
    const char *device_uri,		/* I - device-uri string */
    const char *device_make_and_model,	/* I - device-make-and-model string or @code NULL@ */
    const char *device_info,		/* I - device-info string or @code NULL@ */
    const char *device_id,		/* I - device-id string or @code NULL@ */
    const char *device_location)	/* I - device-location string or @code NULL@ */
{
  if (!device_scheme || !device_uri)
    return;

  printf("%s %s", device_scheme, device_uri);
  if (device_make_and_model && *device_make_and_model)
    quote_string(device_make_and_model);
  else
    quote_string("unknown");
  quote_string(device_info);
  quote_string(device_id);
  quote_string(device_location);
  putchar('\n');
  fflush(stdout);
}


/*
 * 'quote_string()' - Write a quoted string to stdout, escaping \ and ".
 */

static void
quote_string(const char *s)		/* I - String to write */
{
  fputs(" \"", stdout);

  if (s)
  {
    while (*s)
    {
      if (*s == '\\' || *s == '\"')
	putchar('\\');

      if (((*s & 255) < ' ' && *s != '\t') || *s == 0x7f)
        putchar(' ');
      else
        putchar(*s);

      s ++;
    }
  }

  putchar('\"');
}
